state-sync/state-sync-v2/state-sync-driver/src/storage_synchronizer.rs

```diff
@@ -176,3 +230,120 @@ impl<ChunkExecutor: ChunkExecutorTrait> StorageSynchronizerInterface
+                                    &target_ledger_info,
+                                    end_of_epoch_ledger_info.as_ref(),
+                                ).map_err(|error| {
+                                     error!(
+                                        "Failed to execute the transaction data chunk! Error: {:?}", error
+                                    );
+                                    error
+                                })
+                        },
+                        TransactionDataChunk::TransactionOutputs(outputs_with_proof, target_ledger_info, end_of_epoch_ledger_info) => {
+                            chunk_executor
+                                .apply_chunk(
+                                    outputs_with_proof,
+                                    &target_ledger_info,
+                                    end_of_epoch_ledger_info.as_ref(),
+                                ).map_err(|error| {
+                                    error!(
+                                        "Failed to apply the transaction data chunk! Error: {:?}", error
+                                    );
+                                    error
+                                })
+                        }
+                    };
+
+                    // Notify the committer of new executed chunks
+                    if result.is_ok() {
+                        if let Err(error) = committer_notifier.try_send(()) {
+                            error!(
+                                "Failed to notify the committer! Error: {:?}", error
+                            );
+                        }
+                    }
+                }
+            }
+        }
+    };
+
+    // Spawn the executor
+    spawn(runtime, executor);
+}
+
+/// Spawns a dedicated committer that commits executed (but pending) chunks
+fn spawn_committer<ChunkExecutor: ChunkExecutorTrait + 'static>(
+    chunk_executor: Arc<ChunkExecutor>,
+    mut committer_listener: mpsc::Receiver<()>,
+    mut commit_notification_sender: mpsc::UnboundedSender<CommitNotification>,
+    runtime: Option<&Runtime>,
+) {
+    // Create an executor
+    let committer = async move {
+        loop {
+            ::futures::select! {
+                _ = committer_listener.select_next_some() => {
+                    // Commit the executed chunk
+                    match chunk_executor.commit_chunk() {
+                        Ok((events, transactions)) => {
+                            let commit_notification = CommitNotification::new(events, transactions);
+                            if let Err(error) = commit_notification_sender.send(commit_notification).await {
+                                error!("Failed to send commit notification! Error: {:?}", error);
+                            }
+                        }
+                        Err(error) => {
+                            error!("Failed to commit executed chunk! Error: {:?}", error);
```

Comment:
Assuming the error is retry-able, we lost one message in the commit queue here? 
Shall we push the commit command back to the queue? to make the back pressure work?

Reply:
I've added a TODO for this in the code, but my thinking is that if there's any error in the pipeline (executor or committer), I'm going to do the simple thing and force a reset that flushes the pipeline and starts fetching a new data stream from peers. This is ultimately required to handle malicious responses or data streams that are in weird states because of misbehaving peers.

Reply:
That sounds perfect.